File: fair-division-app/Lipton.py
import itertools
import pandas as pd

# Define agents, items, and utilities
agents = ['1', '2', '3']
items = ['o1', 'o2', 'o3', 'o4', 'o5']

utilities = {
    '1': {'o1': 1, 'o2': 2, 'o3': 5, 'o4': 3, 'o5': 7},
    '2': {'o1': 2, 'o2': 6, 'o3': 8, 'o4': 1, 'o5': 5},
    '3': {'o1': 5, 'o2': 4, 'o3': 4, 'o4': 3, 'o5': 12}
}

def bundle_value(agent, bundle):
    return sum(utilities[agent][o] for o in bundle)

def build_envy_graph(allocation):
    graph = {i: [] for i in agents}
    for i, j in itertools.permutations(agents, 2):
        if bundle_value(i, allocation[j]) - bundle_value(i, allocation[i]) > 0:
            graph[i].append(j)
    return graph

def find_cycle(graph):
    # First try to find a 2-cycle (mutual envy)
    for i, j in itertools.permutations(agents, 2):
        if j in graph[i] and i in graph[j]:
            return [i, j]
    # Otherwise use DFS for longer cycles
    visited = set()
    def dfs(u, path):
        visited.add(u)
        path.append(u)
        for v in graph[u]:
            if v in path:
                return path[path.index(v):]
            if v not in visited:
                cycle = dfs(v, path)
                if cycle:
                    return cycle
        path.pop()
        return None

    for a in agents:
        if a not in visited:
            cycle = dfs(a, [])
            if cycle:
                return cycle
    return None

def rotate_cycle(allocation, cycle):
    new_alloc = allocation.copy()
    for idx, agent in enumerate(cycle):
        source = cycle[(idx + 1) % len(cycle)]
        new_alloc[agent] = list(allocation[source])
    return new_alloc

def lipton_allocation(agents, items):
    allocation = {a: [] for a in agents}
    allocation[agents[0]].append(items[0])

    for item in items[1:]:
        # break cycles until none remain
        while True:
            graph = build_envy_graph(allocation)
            cycle = find_cycle(graph)
            if not cycle:
                break
            allocation = rotate_cycle(allocation, cycle)
        # allocate to an agent with no incoming envy
        graph = build_envy_graph(allocation)
        indegree = {a: 0 for a in agents}
        for src, neighs in graph.items():
            for tgt in neighs:
                indegree[tgt] += 1
        for a in agents:
            if indegree[a] == 0:
                allocation[a].append(item)
                break

    # final envy-free rotation
    while True:
        graph = build_envy_graph(allocation)
        cycle = find_cycle(graph)
        if not cycle:
            break
        allocation = rotate_cycle(allocation, cycle)

    return allocation

# Run algorithm on example
alloc = lipton_allocation(agents, items)

# Compute maximum envy and record envy relations
max_envy = 0
envy_relations = []
for i, j in itertools.permutations(agents, 2):
    envy_amount = bundle_value(i, alloc[j]) - bundle_value(i, alloc[i])
    if envy_amount > 0:
        envy_relations.append((i, j, envy_amount))
    max_envy = max(max_envy, envy_amount)

# Display allocation table
df = pd.DataFrame([
    {
        'Agent': a,
        'Bundle': ', '.join(alloc[a]),
        'Utility': bundle_value(a, alloc[a])
    }
    for a in agents
])
print(df.to_string(index=False))

# Print envy relations or "No envy!"
if envy_relations:
    print("\nEnvy relations:")
    for i, j, amt in envy_relations:
        print(f"Agent {i} envies Agent {j} by {amt}")
else:
    print("\nNo envy!")

print(f"\nMaximum envy: {max_envy}")

File: fair-division-app/src/lib/lipton.ts
import type { Preferences, Allocation, EnvyMatrix, UtilityStats } from './picking_sequence';
import { analyzeUtilities, calculateEnvy } from './picking_sequence';

// Helper: compute utility of a bundle for an agent
function bundleValue(agent: string, bundle: string[], prefs: Preferences): number {
	return bundle.reduce((sum, obj) => sum + prefs[agent][obj], 0);
}

// Build envy graph: edges i -> j if i envies j
function buildEnvyGraph(allocation: Allocation, prefs: Preferences): Record<string, string[]> {
	const agents = Object.keys(allocation);
	const graph: Record<string, string[]> = {};
	for (const i of agents) {
		graph[i] = [];
		for (const j of agents) {
			if (i === j) continue;
			const diff =
				bundleValue(i, allocation[j], prefs) - bundleValue(i, allocation[i], prefs);
			if (diff > 0) graph[i].push(j);
		}
	}
	return graph;
}

// Find a cycle, preferring 2-cycles
function findCycle(graph: Record<string, string[]>): string[] | null {
	const agents = Object.keys(graph);
	// check mutual envy (2-cycles)
	for (const i of agents) {
		for (const j of graph[i]) {
			if (graph[j]?.includes(i)) return [i, j];
		}
	}
	// DFS for longer cycles
	const visited = new Set<string>();
	function dfs(u: string, path: string[]): string[] | null {
		visited.add(u);
		path.push(u);
		for (const v of graph[u] || []) {
			const idx = path.indexOf(v);
			if (idx >= 0) return path.slice(idx);
			if (!visited.has(v)) {
				const cycle = dfs(v, path);
				if (cycle) return cycle;
			}
		}
		path.pop();
		return null;
	}
	for (const a of agents) {
		if (!visited.has(a)) {
			const cycle = dfs(a, []);
			if (cycle) return cycle;
		}
	}
	return null;
}

// Rotate bundles along a cycle
function rotateCycle(allocation: Allocation, cycle: string[]): Allocation {
	const newAlloc: Allocation = {};
	for (const a in allocation) newAlloc[a] = [...allocation[a]];
	for (let i = 0; i < cycle.length; i++) {
		const cur = cycle[i];
		const next = cycle[(i + 1) % cycle.length];
		newAlloc[cur] = [...allocation[next]];
	}
	return newAlloc;
}

// Main Lipton algorithm
export function liptonAllocate(prefs: Preferences): Allocation {
	const agents = Object.keys(prefs);
	const objects = Object.keys(prefs[agents[0]]);
	let alloc: Allocation = Object.fromEntries(agents.map((a) => [a, []]));

	// 1) Give first object to first agent
	alloc[agents[0]].push(objects[0]);

	// 2) Sequentially allocate remaining
	for (const o of objects.slice(1)) {
		// break envy cycles
		while (true) {
			const graph = buildEnvyGraph(alloc, prefs);
			const cycle = findCycle(graph);
			if (!cycle) break;
			alloc = rotateCycle(alloc, cycle);
		}
		// pick an agent with no incoming edges
		const graph = buildEnvyGraph(alloc, prefs);
		const indegree: Record<string, number> = Object.fromEntries(agents.map((a) => [a, 0]));
		for (const i of agents) for (const j of graph[i]) indegree[j]++;
		const target = agents.find((a) => indegree[a] === 0)!;
		alloc[target].push(o);
	}

	// 3) Final break of any remaining envy
	while (true) {
		const graph = buildEnvyGraph(alloc, prefs);
		const cycle = findCycle(graph);
		if (!cycle) break;
		alloc = rotateCycle(alloc, cycle);
	}

	return alloc;
}

export function runLipton(prefs: Preferences): {
	allocation: Allocation;
	envyMatrix: EnvyMatrix;
	envyValue: number;
	utilityStats: UtilityStats;
	sequence: number[]; // PS always returns a sequence
} {
	const allocation = liptonAllocate(prefs);

	// compute utilities & envy just like PS
	const utilityStats = analyzeUtilities(allocation, prefs);
	const { envyValue, envyMatrix } = calculateEnvy(allocation, prefs);

	// if not envy‐free, bail out so UI can warn
	if (envyValue > 0) {
		throw new Error('No envy-free allocation possible');
	}

	// Lipton has no “pick” sequence to show, so return empty array
	return { allocation, envyMatrix, envyValue, utilityStats, sequence: [] };
}
